src/hotspot/cpu/s390/c1_LIRAssembler_s390.cpp

```diff
@@ -1533,7 +1533,13 @@ void LIR_Assembler::arith_op(LIR_Code code, LIR_Opr left, LIR_Opr right, LIR_Opr
       jint c = right->as_constant_ptr()->as_jint();
       switch (code) {
         case lir_add: __ z_agfi(lreg, c);  break;
-        case lir_sub: __ z_agfi(lreg, -c); break; // note: -min_jint == min_jint
+        case lir_sub:
+                      if (Immediate::is_simm16(c)) {
+                        __ z_ahi(lreg, -c); // note: -min_jint == min_jint
+                      } else {
+                        __ z_slfi(lreg, c);
+                      }
+                      break;
```

Comment:
Would it be simpler to use `java_negate(c)` (from globalDefinitions.hpp)?

Reply:
Not sure of that actually. I didn't even know that there exists such helper method. Thanks for making me aware. 

I updated current solution in accordance with GCC compiler. So Z don't have a `shi` instruction which can handle 16-bit numbers, so GCC negates the number and adds it with `ahi` instruction. Then for number upto 32bits, `slfi` instruction is emitted for subtraction. 

@RealLucy do you have other thoughts on this ?

Reply:
I agree,
`  __ z_afi(lreg, java_negate(c));`
reads better.

Reply:
Couldn't you use add2reg_32() for the subtraction as well?